### PR TITLE
wait for catalogsource pod deletion

### LINKED ISSUE #2440: "Catalog represents a store of bundles which OLM can use to install Operators config map update triggers registry pod rollout"

### PR DESCRIPTION
**Pull request by akihikokuroda**
Signed-off-by: akihikokuroda <akihikokuroda2020@gmail.com>

**Description of the change:**
Add code that wait for deletion of the catalogsource pod in the CatalogSourceCleanupFunc.

**Motivation for the change:**
Possible fix for issue https://github.com/operator-framework/operator-lifecycle-manager/issues/2440

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:513]: We shouldn't need to nil check the client here, and removing the conditional should improve readability. WDYT?

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:515]: nit: Substitute context.TODO() with context.Background()?

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:517]: We have a dedicated global logger that can be accessed through ctx.Ctx().Logf(...) for any logging purposes.

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:514]: The e2e suite is a bit of a weird mixture of ginkgo/gomega and testify style assertions. Recently, we've been moving to the former and utilizing the `Eventually(...)` closure to handle polling. The equivalent to wait.Poll here would be something like the following:

```go
Eventually(func() (bool, error) {
    fetched, err := ...
    if err != nil {
        return false, err
    }
    if len(fetched.Items) == 0 {
        return true, nil
    }
    return false, nil
}
```

Any thoughts?

edit: I thought more about this, and refactoring this to use the Eventually-style function shouldn't improve too much, so feel free to disregard.

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:518]: nit, feel free to disregard: `return false, err` here as we're technically not ready to exit the poll if we encounter any error while performing list operations.

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:565]: This is an unfortunate side effect of a good chunk of testing helper functions that should be refactored at some point, but we could also just update this overall function to take a `c operatorclient.ClientInterface` parameter, and at the call sites, you can get access to an already initialized client through `newKubeClient()`.

**Review by timflannagan — COMMENTED** [review]
Thanks for trying to tackling this and help cut down on some of the e2e tests that have been failing recently. I had a couple of quick comments/suggestions/etc. inline. I haven't poked around the testing packages to see whether this is an issue with test pollution, e.g. another, unrelated catalogsource test pod is influencing this failing assertion, but introducing additional safeguards should be sufficient. I'm hoping to take a better look later today too.

/ok-to-test

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go:513]: Thanks! I agree. I take out the check.

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go:515]: OK.  I change it.

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go:514]: I leave it for now.

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go:518]: OK.  In case of error, it probably times out.

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go:565]: Thanks.  I change to use newKubeClient() in the helper function instead of taking the client as an argument.

File: test/e2e/util_test.go
Comment (akihikokuroda) [reply test/e2e/util_test.go]: Thanks! I change it.

File: test/e2e/util_test.go
Comment (timflannagan) [reply test/e2e/util_test.go]: Sounds good - we'll eventually need to refactor those helper function signatures, but this should (hopefully) cut down on e2e flakes in the interim.

**Comment by timflannagan** [comment]
I think these changes look reasonable to me, so I'll preemptively tag it while we wait for e2e run - we'll have to chase down an approver next week before it can merge.

/lgtm

**Comment by akihikokuroda** [comment]
How can I rerun the failing test?

**Comment by timflannagan** [comment]
@akihikokuroda I'm not entirely sure what kind of repository permissions you would need to restart checks unfortunately - I'm going manually recycle them as the workflows in this PR have been renamed.

File: test/e2e/util_test.go
Comment (timflannagan) [inline test/e2e/util_test.go:523]: Just coming back to this:

- It doesn't look like we're handling the wait.Poll error anywhere.
- We should try and aim for a better logging as well

WDYT about the following:

```go
Eventually(func() (bool, error) {
	fetched, err := newKubeClient().KubernetesInterface().CoreV1().Pods(catalogSource.GetNamespace()).List(context.Background(), metav1.ListOptions{LabelSelector: "olm.catalogSource=" + catalogSource.GetName()})
	if err != nil {
		return false, err
	}
	if len(fetched.Items) == 0 {
		return true, nil
	}
	ctx.Ctx().Logf("waiting for the catalog source %s pod to be deleted...", fetched.GetName())
	return false, nil
}).Should(BeTrue())
```

Comment (akihikokuroda) [reply test/e2e/util_test.go:523]: Thanks!  I'll take your changes.

File: test/e2e/util.go
Comment (timflannagan) [inline test/e2e/util.go:510]: Any thoughts on whether we should update the parent function signature and pass the kube client as a parameter vs. utilizing the helper function to get it at runtime? I'm not super I have a strong opinion, but wanted to get another opinion if we think passing it as a parameter is cleaner (or easier to maintain).

Comment (akihikokuroda) [reply test/e2e/util.go:510]: I don't have a strong opinion, either.  I like this way because it makes this PR changes small.  I can implement passing the kube client as an argument.  It needs some additional changes in this file and one more file.  What do you think?

Comment (timflannagan) [reply test/e2e/util.go:510]: Sounds reasonable to me - lets hold off.

/lgtm

Comment (awgreene) [reply test/e2e/util.go:510]: I think we should fix this holistically across the test suite, I've created [this issue](https://github.com/operator-framework/operator-lifecycle-manager/issues/2570) to track the work.

**Review by awgreene — APPROVED** [review]
Great work.
/approve

**Comment by timflannagan** [comment]
@akihikokuroda Can you rebase this when you get a chance? I think this is ready to go but I don't want to tag this PR with a lgtm and have the openshift-bot spam retest commands in the meantime.

**Review by timflannagan — APPROVED** [review]
Thanks!

/lgtm

**Pull request merged**
